specs/emip-005.md

```diff
@@ -0,0 +1,472 @@
+COSE_recipient = [
+  Headers,
+  ciphertext : bstr / nil, # contains encrypted key to decrypt the COSE_Encrypt ciphertext
+  ? recipients : [+COSE_recipient] # in case you need multiple rounds to get decryption key
+]
+```
+
+To encrypt, the structure as a whole, we call our encryption method once for each level (root, recipient, etc.) recursively. For example, we encrypt the `signed_message` and put it in the `COSE_Encrypt` ciphertext, then we encrypt the decryption key and put it in the `COSE_recipient` ciphertext.
+
+For the `Headers`,
+
+- The `protected` fields MUST be empty. These are meant to be used with AEAD which we  don't need in this  specification (you can read more about it at [RFC 8152 section 5.3](https://tools.ietf.org/html/rfc8152#section-5.3)).
+- The `unprotected` field MUST contain a `enc_type` label with value `EncryptionType` so that wallets can easily know what UI to display to users.
+
+We define the encrypted payload as the union of two possibilities (defined below)
+
+```
+encrypted_message = PasswordEncryption / PubKeyEncryption
+```
+
+#### Password-based encryption
+
+For password-based encryption we don't need a receiver field (anybody who knows the password can decrypt) so we instead use the following (simplified) structure
+
+```
+COSE_Encrypt0 = [
+    Headers,
+    ciphertext : bstr / nil,
+]
+
+PasswordEncryption = 16 (COSE_Encrypt0)
+```
+
+The `Headers` MUST have `enc_type: 1` in the `unprotected` field.
+
+The `ciphertext` is simply the result of [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md) on the `signed_message`.
+
+#### Public key based encryption
+
+We only allow encrypting based on `ED25519 extended` public keys (the ones used for Cardano). To encrypt based on these public keys, you must
+
+1) Compute a password consisting of 22 case-sensitive alphanumeric (a–z, A–Z, 0–9) characters (this gives you ~128 bits of entropy)
+
+Now, for each receiver, you must
+
+2) Compute an ephemeral key pair using `ED255519 extended`
+1) Compute the shared DH secret between the private key from step (1) and the public key received (using the `exchange` functionality)
+1) Compute the `Blake2b512` hash of the secret
+1) Use this as the password to encrypt the password in (1) using [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md)
+
+The structure will look like the following:
+
+```
+COSE_Encrypt = [
+  Headers,
+  ciphertext : bstr / nil, # contained signed_message encrypted with random password
+  recipients : [+COSE_recipient]
+]
+
+COSE_recipient = [
+  Headers,
+  ciphertext : bstr / nil, # contains random password encrypted with shared secret
+]
+
+PubKeyEncryption = 96 (COSE_Encrypt)
+```
+
+The `Headers` for the body MUST have `enc_type: 2` in the `unprotected` field.
```

Comment:
Let's suppose, A sends (Public key based) encrypted memo to B. In this case `enc_type` would be `2`. What if man-in-the-middle adversary C: 1) modifies `enc_type` (it's unprotected) from `2` to `1` in `COSE_Encrypt` structure, 2) modifies CBOR tag from `16` to `96` (this is tag of `COSE_Encrypt0` structure - Table 1 in https://tools.ietf.org/html/rfc8152) and 3) removes `COSE_recipient` structure (that contains encrypted password and `eph_pub`) from message? In this case B's wallet will prompt B to enter password and try to decrypt the memo with password-based decryption algorithm. Since passwords won't match, B is unable to decrypt memo, and doesn't understand why. 

It's not a security issue, but seems this case is a little bit worse than detectable active man-in-the-middle, since C removes unobserved.